File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        res = []
        for i in s:
            if i == "(" or i == "{" or i =="[":
                res.append(i)
            if i == ")":
                if not res:
                    return False
                if res.pop() != "(":
                    return False

            if i == "}":
                if not res:
                    return False
                if res.pop() != "{":
                    return False
            
            if i == "]":
                if not res:
                    return False
                if res.pop() != "[":
                    return False
        if res:
            return False
        return True
            
            
            

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Map<Character, Character> map = new HashMap<>();
        map.put(')','(');
        map.put(']','[');
        map.put('}','{');
        Stack<Character> stck = new Stack<>();

        for(char i: s.toCharArray()) {
            
            if(i == '(' || i == '{' || i == '[') {
                stck.push(i);
            }
            else {
                if(stck.isEmpty()) return false;
                else if(map.get(i) != stck.peek()) return false;
                else stck.pop();
            }
        }

        if(!stck.isEmpty()) return false;
        return true;
        
    }
}